packages/block-editor/src/components/block-draggable/index.js

```diff
@@ -61,6 +78,55 @@ const BlockDraggable = ( {
 		};
 	}, [] );
 
+	// Find the root of the editor iframe.
+	const blockRef = useBlockRef( clientIds[ 0 ] );
+	const editorRoot = blockRef.current?.closest( 'body' );
+
+	// Add a dragover event listener to the editor root to track the blocks being dragged over.
+	// The listener has to be inside the editor iframe otherwise the target isn't accessible.
+	// Check if the dragged blocks are allowed inside the target. If not, grey out the draggable.
+	useEffect( () => {
+		if ( ! editorRoot ) {
+			return;
+		}
+
+		const onDragOver = ( event ) => {
+			if ( ! event.target.closest( '[data-block]' ) ) {
+				return;
+			}
+
+			const newTargetClientId = event.target
+				.closest( '[data-block]' )
+				.getAttribute( 'data-block' );
+			//Only update targetClientId if it has changed
+			// and if it's a container block
+			if (
+				targetClientId !== newTargetClientId &&
+				getBlockListSettings( newTargetClientId )
```

Comment:
Is `getBlockListSettings( newTargetClientId )` checking for nested blocks to see if it's a container?

I think calling a select outside `useSelect` is an anti-pattern as it risks not returning fresh data when the store is updated.

I tested by updating the list settings for a block via `wp.data.dispatch('core/block-editor').updateBlockListSettings()` and it looks like the return state is updated, I guess because a rerender is triggered by another prop. So it's working in that case as far as I can tell.

cc @jsnajdr in case I missed something.

Reply:
I don't _think_ it's an anti-pattern, we use select functions outside of `useSelect` all over the place 😅 

`getBlockListSettings` was the most reliable way I found of checking whether a block is a container, given that only containers are block lists (and they might not have any inner blocks if they're empty). I'm definitely open to ideas on this one though 😄

Reply:
Actually scratch that, I don't think we need to check for the container anymore 😂 now we're using the inserter visibility as per @andrewserong's suggestion. 

Still keen to hear about preferred code patterns though 👀

Reply:
Here `getBlockListSettings` is called in an event handler (`onDragOver`) and that's fine. We want to read the fresh value when the event is being dispatched, and that's what happens here. No rerenders need to be triggered when the value changes.

The only thing I would change here is to write the code, for clarity, as two separate `useSelect` calls:
```js
const { getBlockListSettings } = useSelect( blockEditorStore );
// or alternatively
const { getBlockListSettings } = useRegistry().select( blockEditorStore );

const { ...rest } = useSelect( ( select ) => ... );
```
One of them just gets a reference to the selector, without any store subscriptions or reactivity, and the second reads data in the usual reactive way.

Reply:
Thanks for confirming and setting me straight on that one, @jsnajdr!

Reply:
Ok thanks for the clarification!

Reply:
> The only thing I would change here is to write the code, for clarity, as two separate useSelect calls

There are other selectors we're grabbing directly and using inside the effect such as getBlockType and getAllowedBlocks. I guess following that logic they should all be in a separate useSelect call?